src/Control/Effect/Cut.hs

```diff
@@ -0,0 +1,69 @@
+{-# LANGUAGE DeriveFunctor, ExistentialQuantification, FlexibleContexts, FlexibleInstances, LambdaCase, MultiParamTypeClasses, StandaloneDeriving, TypeOperators, UndecidableInstances #-}
+module Control.Effect.Cut
+( Cut(..)
+, cutfail
+, call
+, cut
+, Branch(..)
+, branch
+, runCut
+, CutC(..)
+) where
+
+import Control.Applicative (Alternative(..))
+import Control.Effect.Carrier
+import Control.Effect.Internal
+import Control.Effect.NonDet
+import Control.Effect.Sum
+
+data Cut m k
```

Comment:
Just to ensure I'm understanding correctly, the `Cut` effect is described in Effect Handlers in Scope (even though our handler doesn't look anything like the one in literature). `Cut` has thse two operations: 
1. `Cutfail`, a hard fail
2. `Call`, which delimits the scope _where_ `Cutfail` should fail within

`Cut` is committed choice, and the entire purpose is so that once we've matched the branch we want, we don't try any other branch. This is a useful capability since we don't have to backtrack (backtracking is quadratic). 

You want to do different thing when you `Fail` vs. `Cutfail` because `Cutfail` is when you stop pursuing other branches. 

@robrix is this why we don't do the table lookup in our other branch? Because we're getting rid of the need to do so (which requires maintaining a representation of trees) or am I getting things mixed up?

Reply:
You’ve got it 👍

> @robrix is this why we don't do the table lookup in our other branch? Because we're getting rid of the need to do so (which requires maintaining a representation of trees) or am I getting things mixed up?

The way we were doing the table-based choice thing depended on having a representation of the computation as a tree that it could inspect, yes. It’s not clear how to do that with `fused-effects`, but it’s likely possible to recover that behaviour with some creativity. But I’m hoping that the efficiency gains from using `fused-effects` in the first place will outweigh the cost of not doing the efficient table-based choice, and therefore that we won’t have a pressing need to reimplement tabling.